include/boost/geometry/algorithms/detail/buffer/piece_border.hpp

```diff
@@ -0,0 +1,509 @@
+            return result;
+        }
+        std::copy(m_ring->begin() + m_begin,
+                  m_ring->begin() + m_end,
+                  std::back_inserter(result));
+        std::copy(m_originals.begin(),
+                  m_originals.begin() + m_original_size,
+                  std::back_inserter(result));
+        // Add the closing point
+        result.push_back(*(m_ring->begin() + m_begin));
+
+        return result;
+    }
+
+    void get_properties_of_border(bool is_point_buffer, Point const& center)
+    {
+        m_has_envelope = calculate_envelope(m_envelope);
+        if (m_has_envelope)
+        {
+            // Take roundings into account, enlarge box
+            detail::buffer::buffer_box(m_envelope, 1.0e-3, m_envelope);
+        }
+        if (valid() && is_point_buffer)
+        {
+            // Determine min/max radius
+            calculate_radii(center, m_ring->begin() + m_begin, m_ring->begin() + m_end);
+        }
+    }
+
+    template <typename SideStrategy>
+    void get_properties_of_offsetted_ring_part(SideStrategy const& strategy)
+    {
+        if (valid())
+        {
+            m_is_convex = is_convex(strategy);
+            check_monotonicity(m_ring->begin() + m_begin, m_ring->begin() + m_end);
+        }
+    }
+
+    void set_offsetted(Ring const& ring, std::size_t begin, std::size_t end)
+    {
+        BOOST_GEOMETRY_ASSERT(begin <= end);
+        BOOST_GEOMETRY_ASSERT(begin < boost::size(ring));
+        BOOST_GEOMETRY_ASSERT(end <= boost::size(ring));
+
+        m_ring = &ring;
+        m_begin = begin;
+        m_end = end;
+    }
+
+    void add_original_point(Point const& point)
+    {
+        BOOST_GEOMETRY_ASSERT(m_original_size < 2);
+        m_originals[m_original_size++] = point;
+    }
+
+    template <typename Box>
+    bool calculate_envelope(Box& envelope) const
+    {
+        geometry::assign_inverse(envelope);
+        if (! valid())
+        {
+            return false;
+        }
+        expand_envelope(envelope, m_ring->begin() + m_begin, m_ring->begin() + m_end);
+        expand_envelope(envelope, m_originals.begin(), m_originals.begin() + m_original_size);
+        return true;
+    }
+
+
+    // Whatever the return value, the state should be checked.
+    template <typename TurnPoint, typename State>
+    bool point_on_piece(TurnPoint const& point,
+                        bool one_sided, bool is_linear_end_point,
+                        State& state) const
+    {
+        if (! valid())
+        {
+            return false;
+        }
+
+        // Walk over the different parts of the ring, in clockwise order
+        // For performance reasons: start with the helper part (one segment)
+        // then the original part (one segment, if any), then the other helper
+        // part (one segment), and only then the offsetted part
+        // (probably more segments, check monotonicity)
+        geometry::strategy::buffer::turn_in_ring_winding<coordinate_type> tir;
```

Comment:
Right, so this is a cartesian strategy right? I assume it is and in the future it should be derived from the strategy passed by the user because in general should depend on the CS. So in geographic CS the ellipsoid should be propagated here. Could we at for now implement it in such a way that when someone tries to pass a non-cartesian geometry it does not compile? E.g. add static assert checking cs_tag == cartesian_tag or something. This way it will be easier to find it in the future. It might be done here or in the strategy itself.

Reply:
Actually, the strategy passed by the user should be checked to be cartesian strategy, not the geometry. The intersection strategy defines cs_tag AFAIR so this should be easy.

Reply:
I leave this for now, first it was already targetted to cartesian (I made it less dependent by removing side_by_intersection and other specific things, but this PR does not have the purpose to make it completely cs agnostic). Second, buffer of geographic points is already supported so I should not blindly check the cs_tag.

Reply:
> this PR does not have the purpose to make it completely cs agnostic

Right.

> Second, buffer of geographic points is already supported so I should not blindly check the cs_tag.

Ok, fixing it now would probably be outside of this PR.

Also my thoughts regarding this issue. Internally `turn_in_ring_winding` relies on cartesian `model::infinite_line` (unless I'm missing something). It is therefore a cartesian strategy. Using it in spherical and geographic buffer is technically wrong (unless you're certain that the distances passed by the user are small which we can't guarantee). Checking it wouldn't be blind, it'd make sure that the user doesn't get the incorrect results thinking they are correct. I'd rather argue that using it in the first place for all CSes was blindly assuming things. If it is used for geographic points then it means it's a bug. It should probably have been modified along with introducing support for geographic points. My guess is that it was not because the compiler didin't argue about it, if the CS was checked here the compiler would point to the place that is wrong in the code.

Reply:
Any buffer around a single point will not call this code.
Only when multi-points (or linear/poly geometries) are buffered, then this code would be called.
But if we make a compile-time check here, the geo-point-buffer won't compile. That is the reason I cannot do it now. It was already like this before (but then not `infinite_line` (indeed) but `side_by_intersection` and other things which were cartesian related)

Reply:
Ok, got it, no problem. As I said, fixing it is outside of the scope of this PR.